File: sheets.js
'use strict';

// sheets.js — port dari sheets.go + teams.go. Export Blast Log ke Google Sheets via
// service account, ensureSheetExists, dan sync mapping team dari tab "Query Blast".

const fs = require('fs');
const { google } = require('googleapis');
const { db } = require('./db');
const { normalizePhone } = require('./util');

let sheetsSvc = null;
let spreadsheetID = '';
let sheetName = 'Blast Log';

async function initSheets() {
  const saPath = process.env.GOOGLE_SERVICE_ACCOUNT_JSON;
  spreadsheetID = process.env.GSHEET_SPREADSHEET_ID || '';
  sheetName = process.env.GSHEET_SHEET_NAME || 'Blast Log';

  if (!saPath || !spreadsheetID) {
    console.log('Sheets export disabled (set GOOGLE_SERVICE_ACCOUNT_JSON + GSHEET_SPREADSHEET_ID di .env untuk aktifkan)');
    return;
  }
  if (!fs.existsSync(saPath)) {
    throw new Error(`service account JSON tidak ditemukan di ${saPath}`);
  }
  const auth = new google.auth.GoogleAuth({
    keyFile: saPath,
    scopes: ['https://www.googleapis.com/auth/spreadsheets'],
  });
  sheetsSvc = google.sheets({ version: 'v4', auth });
  console.log('Sheets export enabled — spreadsheet:', spreadsheetID, '/ sheet:', sheetName);
}

function enabled() {
  return sheetsSvc !== null;
}
function getSpreadsheetID() {
  return spreadsheetID;
}
function sheetURL() {
  return `https://docs.google.com/spreadsheets/d/${spreadsheetID}/edit`;
}
function reportSheetName() {
  return process.env.GSHEET_REPORT_SHEET_NAME || 'Belum Respons';
}
function querySheetName() {
  return process.env.GSHEET_QUERY_SHEET_NAME || 'Query Blast';
}

async function ensureSheetExists(name) {
  const ss = await sheetsSvc.spreadsheets.get({ spreadsheetId: spreadsheetID });
  for (const s of ss.data.sheets || []) {
    if (s.properties && s.properties.title === name) return;
  }
  await sheetsSvc.spreadsheets.batchUpdate({
    spreadsheetId: spreadsheetID,
    requestBody: { requests: [{ addSheet: { properties: { title: name } } }] },
  });
}

async function clearAndWrite(name, clearCols, values) {
  await sheetsSvc.spreadsheets.values.clear({ spreadsheetId: spreadsheetID, range: `${name}!${clearCols}` });
  await sheetsSvc.spreadsheets.values.update({
    spreadsheetId: spreadsheetID,
    range: `${name}!A1`,
    valueInputOption: 'USER_ENTERED',
    requestBody: { values },
  });
}

async function handleExportSheet(req, res) {
  if (!sheetsSvc) {
    return res.status(400).send('Sheets export belum dikonfigurasi. Set GOOGLE_SERVICE_ACCOUNT_JSON + GSHEET_SPREADSHEET_ID di .env, lalu restart.');
  }
  const rows = db
    .prepare(
      `SELECT b.started_at, COALESCE(b.user_name,'') AS user_name, b.user_email, r.phone,
              COALESCE(r.nama_outlet,'') AS nama_outlet, COALESCE(r.nomer_invoice,'') AS nomer_invoice,
              r.status, COALESCE(r.error,'') AS error, COALESCE(r.sent_at,'') AS sent_at,
              COALESCE(r.message,'') AS message, r.blast_log_id
       FROM blast_recipients r JOIN blast_logs b ON r.blast_log_id = b.id ORDER BY r.id ASC`
    )
    .all();

  const values = [
    ['Waktu Blast', 'User', 'Email', 'Phone', 'Nama Outlet', 'Nomor Invoice', 'Status', 'Error', 'Sent At', 'Pesan', 'Blast ID'],
  ];
  for (const r of rows) {
    values.push([
      r.started_at, r.user_name, r.user_email, "'" + r.phone, r.nama_outlet, r.nomer_invoice,
      r.status, r.error, r.sent_at, r.message, r.blast_log_id,
    ]);
  }

  try {
    await clearAndWrite(sheetName, 'A:K', values);
  } catch (e) {
    return res.status(500).send('export sheet: ' + e.message + '. Pastikan service account punya akses Editor.');
  }
  res.json({ ok: true, rows: rows.length, sheet_url: sheetURL(), sheet_name: sheetName });
}

function handleSheetStatus(req, res) {
  res.json({ enabled: sheetsSvc !== null, spreadsheet: spreadsheetID, sheet_name: sheetName, sheet_url: sheetURL() });
}

// syncTeamsFromSheet — baca tab "Query Blast", update kolom team/area di chat_threads
// dengan cocokkan by phone (62...). Return jumlah thread ter-update.
async function syncTeamsFromSheet() {
  if (!sheetsSvc) throw new Error('Sheets belum dikonfigurasi (GOOGLE_SERVICE_ACCOUNT_JSON + GSHEET_SPREADSHEET_ID)');
  const sn = querySheetName();
  const resp = await sheetsSvc.spreadsheets.values.get({ spreadsheetId: spreadsheetID, range: `${sn}!A:Z` });
  const vals = resp.data.values || [];
  if (vals.length < 2) throw new Error(`tab '${sn}' kosong / tidak ada baris data`);

  const idx = {};
  vals[0].forEach((c, i) => (idx[String(c).trim().toLowerCase()] = i));
  if (!('phone' in idx) || !('team' in idx)) {
    throw new Error(`kolom 'Phone' & 'Team' wajib ada di tab '${sn}'`);
  }
  const cell = (row, i) => (i >= 0 && i < row.length ? String(row[i]).trim() : '');

  const map = new Map(); // phone -> {team, area}; baris terakhir menang
  for (let r = 1; r < vals.length; r++) {
    const row = vals[r];
    const phone = normalizePhone(cell(row, idx.phone));
    const team = cell(row, idx.team);
    if (!phone || !team) continue;
    const area = 'area' in idx ? cell(row, idx.area) : '';
    map.set(phone, { team, area });
  }

  const upd = db.prepare(`UPDATE chat_threads SET team = ?, area = ? WHERE phone = ?`);
  let updated = 0;
  for (const [phone, v] of map) {
    const info = upd.run(v.team, v.area, phone);
    if (info.changes > 0) updated++;
  }
  return updated;
}

async function handleSyncTeams(req, res) {
  try {
    const n = await syncTeamsFromSheet();
    res.json({ ok: true, updated: n, sheet: querySheetName() });
  } catch (e) {
    res.status(500).send(e.message);
  }
}

module.exports = {
  initSheets,
  enabled,
  getSpreadsheetID,
  sheetURL,
  ensureSheetExists,
  clearAndWrite,
  reportSheetName,
  handleExportSheet,
  handleSheetStatus,
  handleSyncTeams,
};

File: sheets.go
package main

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"time"

	"google.golang.org/api/option"
	"google.golang.org/api/sheets/v4"
)

var (
	sheetsSvc     *sheets.Service
	spreadsheetID string
	sheetName     string
	sheetSALabel  string
)

// initSheets — opsional. Kalau env var tidak diset, fitur export ke Sheets dinonaktifkan
// (endpoint /api/export-sheet akan return error message yang menjelaskan).
func initSheets() error {
	saPath := os.Getenv("GOOGLE_SERVICE_ACCOUNT_JSON")
	spreadsheetID = os.Getenv("GSHEET_SPREADSHEET_ID")
	sheetName = os.Getenv("GSHEET_SHEET_NAME")
	if sheetName == "" {
		sheetName = "Blast Log"
	}

	if saPath == "" || spreadsheetID == "" {
		fmt.Println("Sheets export disabled (set GOOGLE_SERVICE_ACCOUNT_JSON + GSHEET_SPREADSHEET_ID di .env untuk aktifkan)")
		return nil
	}

	if _, err := os.Stat(saPath); err != nil {
		return fmt.Errorf("service account JSON tidak ditemukan di %s: %w", saPath, err)
	}

	ctx := context.Background()
	svc, err := sheets.NewService(ctx, option.WithCredentialsFile(saPath))
	if err != nil {
		return fmt.Errorf("init sheets client: %w", err)
	}
	sheetsSvc = svc
	sheetSALabel = saPath
	fmt.Println("Sheets export enabled — spreadsheet:", spreadsheetID, "/ sheet:", sheetName)
	return nil
}

func handleExportSheet(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "POST only", 405)
		return
	}
	if sheetsSvc == nil {
		httpErr(w, 400, "Sheets export belum dikonfigurasi. Set GOOGLE_SERVICE_ACCOUNT_JSON + GSHEET_SPREADSHEET_ID di .env, lalu restart backend.")
		return
	}

	rows, err := auditDB.Query(`
SELECT
	b.started_at,
	COALESCE(b.user_name, ''),
	b.user_email,
	r.phone,
	COALESCE(r.nama_outlet, ''),
	COALESCE(r.nomer_invoice, ''),
	r.status,
	COALESCE(r.error, ''),
	COALESCE(r.sent_at, ''),
	COALESCE(r.message, ''),
	r.blast_log_id,
	COALESCE(b.attempt, 1)
FROM blast_recipients r
JOIN blast_logs b ON r.blast_log_id = b.id
ORDER BY r.id ASC`)
	if err != nil {
		httpErr(w, 500, "query: %v", err)
		return
	}
	defer rows.Close()

	values := [][]any{{
		"Waktu Blast",
		"User",
		"Email",
		"Attempt",
		"Phone",
		"Nama Outlet",
		"Nomor Invoice",
		"Status",
		"Error",
		"Sent At",
		"Pesan",
		"Blast ID",
	}}

	count := 0
	for rows.Next() {
		var startedAt, userName, userEmail, phone, namaOutlet, nomerInv, status, errMsg, sentAt, message string
		var blastID int64
		var attempt int
		if err := rows.Scan(&startedAt, &userName, &userEmail, &phone, &namaOutlet, &nomerInv, &status, &errMsg, &sentAt, &message, &blastID, &attempt); err != nil {
			continue
		}
		// Excel/Sheets butuh phone sebagai string supaya tidak di-convert ke scientific notation
		// → prefix ' agar Sheets treat as text.
		values = append(values, []any{
			startedAt, userName, userEmail, attempt, "'" + phone, namaOutlet, nomerInv, status, errMsg, sentAt, message, blastID,
		})
		count++
	}

	ctx, cancel := context.WithTimeout(r.Context(), 30*time.Second)
	defer cancel()

	// Clear sheet (range A:L cover semua kolom — termasuk Attempt)
	clearRange := fmt.Sprintf("%s!A:L", sheetName)
	if _, err := sheetsSvc.Spreadsheets.Values.Clear(spreadsheetID, clearRange, &sheets.ClearValuesRequest{}).Context(ctx).Do(); err != nil {
		httpErr(w, 500, "clear sheet: %v. Pastikan service account punya akses Editor ke spreadsheet.", err)
		return
	}

	// Write all values
	writeRange := fmt.Sprintf("%s!A1", sheetName)
	if _, err := sheetsSvc.Spreadsheets.Values.Update(spreadsheetID, writeRange, &sheets.ValueRange{Values: values}).
		ValueInputOption("USER_ENTERED").
		Context(ctx).Do(); err != nil {
		httpErr(w, 500, "write sheet: %v", err)
		return
	}

	sheetURL := fmt.Sprintf("https://docs.google.com/spreadsheets/d/%s/edit", spreadsheetID)
	writeJSON(w, map[string]any{
		"ok":         true,
		"rows":       count,
		"sheet_url":  sheetURL,
		"sheet_name": sheetName,
	})
}

// ensureSheetExists — bikin tab (sheet) baru kalau belum ada. Dipakai report export
// supaya tab "Belum Respons" otomatis dibuat tanpa user harus bikin manual.
func ensureSheetExists(ctx context.Context, name string) error {
	ss, err := sheetsSvc.Spreadsheets.Get(spreadsheetID).Context(ctx).Do()
	if err != nil {
		return fmt.Errorf("get spreadsheet: %w", err)
	}
	for _, s := range ss.Sheets {
		if s.Properties != nil && s.Properties.Title == name {
			return nil
		}
	}
	_, err = sheetsSvc.Spreadsheets.BatchUpdate(spreadsheetID, &sheets.BatchUpdateSpreadsheetRequest{
		Requests: []*sheets.Request{{
			AddSheet: &sheets.AddSheetRequest{Properties: &sheets.SheetProperties{Title: name}},
		}},
	}).Context(ctx).Do()
	return err
}

// handleSheetStatus — quick check apakah Sheets export aktif (untuk UI button enable/disable)
func handleSheetStatus(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, map[string]any{
		"enabled":     sheetsSvc != nil,
		"spreadsheet": spreadsheetID,
		"sheet_name":  sheetName,
		"sheet_url":   fmt.Sprintf("https://docs.google.com/spreadsheets/d/%s/edit", spreadsheetID),
	})
}
